File: core/sidecar/src/main/kotlin/com/circuitstitch/deferno/core/sidecar/SidecarProtocol.kt
package com.circuitstitch.deferno.core.sidecar

/**
 * Stable names for the Sidecar [[Sidecar protocol]] — the protocol version, the seed methods/topics,
 * and the advertised capability ids. Centralised here so the client, the stub Helper, the tests, and
 * the `contracts/sidecar/` spec all agree on the exact strings.
 *
 * The two seed methods exist to prove the contract's three traffic shapes end-to-end (#118); the real
 * capability surface grows over the *same* contract in #119/#120/#123+ (ADR-0024 — "additive over the
 * same socket contract").
 */

/** The protocol version this client speaks; bumped only on a breaking wire change. Sent in [SidecarFrame.Hello]. */
const val SIDECAR_PROTOCOL_VERSION: Int = 1

/** Method names a client sends in a [SidecarFrame.Request]. */
object SidecarMethods {
    /** Request/response: query one capability's current permission state → [PermissionStatusWire]. */
    const val QueryPermission: String = "queryPermission"

    /**
     * Request/response: resolve one capability's permission *without* engaging the capability (#120) —
     * prompts iff `not_determined`, answers the settled [PermissionStatusWire], pushes
     * [SidecarTopics.PermissionChanged] as it settles.
     */
    const val RequestPermission: String = "requestPermission"

    /** Server stream: subscribe to on-device dictation; yields [TranscriptWire] events until cancelled. */
    const val SubscribeTranscript: String = "subscribeTranscript"

    /** Request/response: post a user-visible OS notification ([PostNotificationWire] params, #123). */
    const val PostNotification: String = "postNotification"

    /** Request/response: show/hide the Helper's menu-bar status item ([SetStatusItemWire] params, #125). */
    const val SetStatusItem: String = "setStatusItem"

    /** Request/response: register a global hotkey ([RegisterHotkeyWire] params; replaces its id, #125). */
    const val RegisterHotkey: String = "registerHotkey"

    /** Request/response: unregister a global hotkey ([UnregisterHotkeyWire] params; idempotent, #125). */
    const val UnregisterHotkey: String = "unregisterHotkey"
}

/** Topics a Helper sends in an unsolicited [SidecarFrame.Push]. */
object SidecarTopics {
    /** A capability's permission state changed out-of-band (payload: [PermissionStatusWire]). */
    const val PermissionChanged: String = "permissionChanged"

    /** The visible menu-bar status item was clicked (payload: `{}`, #125). */
    const val StatusItemClicked: String = "statusItemClicked"

    /** A registered global hotkey fired (payload: [HotkeyFiredWire], #125). */
    const val HotkeyFired: String = "hotkeyFired"
}

/** Capability ids a Helper advertises in [SidecarFrame.Welcome] (D4). */
object SidecarCapabilities {
    /** The Helper can answer [SidecarMethods.QueryPermission] and emit [SidecarTopics.PermissionChanged]. */
    const val Permissions: String = "permissions"

    /** The Helper hosts an on-device speech engine reachable via [SidecarMethods.SubscribeTranscript]. */
    const val SpeechTranscribe: String = "speech.transcribe"

    /** The Helper can deliver OS notifications via [SidecarMethods.PostNotification] (#123). */
    const val Notifications: String = "notifications"

    /** The Helper can host a menu-bar status item ([SidecarMethods.SetStatusItem], #125). */
    const val StatusItem: String = "statusItem"

    /** The Helper can register global hotkeys ([SidecarMethods.RegisterHotkey], #125). */
    const val Hotkeys: String = "hotkeys"
}

/**
 * The capability ids carried inside [PermissionStatusWire.capability] / [QueryPermissionWire.capability]
 * (the Swift Helper's `SidecarPermissionCapability`) — which *permission* a status is about, distinct
 * from the [SidecarCapabilities] a Helper advertises.
 */
object SidecarPermissionCapabilities {
    const val Microphone: String = "mic"
    const val Speech: String = "speech"
    const val Notifications: String = "notifications"
}

File: helpers/macos/Sources/SidecarKit/Protocol/SidecarProtocol.swift
import Foundation

/// Stable names for the Sidecar protocol — the Swift mirror of `core/sidecar`'s `SidecarProtocol.kt`
/// and `contracts/sidecar/protocol-v1.md`. The JVM client and this helper are two implementations of
/// **one** contract; these constants must agree string-for-string.

/// The protocol version this helper speaks; sent in `welcome`. Bumped only on a breaking wire change.
public let sidecarProtocolVersion: Int = 1

/// 1 MiB — the frame size cap (`contracts/sidecar/protocol-v1.md`); an over-cap length is a protocol
/// error, never an allocation.
public let sidecarMaxFrameBytes: Int = 1 * 1024 * 1024

/// Method names a client sends in a `request`.
public enum SidecarMethods {
    /// Request/response: query one capability's current permission state → `PermissionStatus`.
    public static let queryPermission = "queryPermission"
    /// Server stream: subscribe to on-device dictation; yields `TranscriptEvent`s until cancelled.
    public static let subscribeTranscript = "subscribeTranscript"
    /// Request/response: post a user-visible OS notification (`PostNotificationRequest` params, #123).
    public static let postNotification = "postNotification"
}

/// Topics the helper sends in an unsolicited `push`.
public enum SidecarTopics {
    /// A capability's permission state changed out-of-band (payload: `PermissionStatus`).
    public static let permissionChanged = "permissionChanged"
}

/// Capability ids the helper advertises in `welcome`.
public enum SidecarCapabilities {
    /// Can answer `queryPermission` and emit `permissionChanged`.
    public static let permissions = "permissions"
    /// Hosts an on-device speech engine reachable via `subscribeTranscript`.
    public static let speechTranscribe = "speech.transcribe"
    /// Can deliver OS notifications via `postNotification` (#123).
    public static let notifications = "notifications"
}

/// The capability ids used inside a `PermissionStatus.capability` field (ADR-0024).
public enum SidecarPermissionCapability {
    public static let microphone = "mic"
    public static let speech = "speech"
    public static let notifications = "notifications"
}
